nltk/tag/crfsuite.py

```diff
@@ -0,0 +1,162 @@
+             - Is Capitalized ?
+             - Has Punctuation ?
+             - Has Number ?
+             - Suffixes up to length 3 
+        :return : a string which contains the features   
+        
+        """ 
+        feature_list = []  
+        # Capitalization 
+        if data[0].isupper():
+            feature_list.append('CAPITALIZATION')
+        
+        # Number 
+        pattern = re.compile('\\d')
+        if re.search(pattern, data) is not None:
+            feature_list.append('HAS_NUM') 
+        
+        # Punctuation
+        punc_cat = set(["Pc", "Pd", "Ps", "Pe", "Pi", "Pf", "Po"])
+        if all (unicodedata.category(x) in punc_cat for x in data):
+            feature_list.append('PUNCTUATION')
+        
+        # Suffix up to length 3
+        if len(data) > 1:
+            feature_list.append('SUF_' + data[-1:]) 
+        if len(data) > 2: 
+            feature_list.append('SUF_' + data[-2:])    
+        if len(data) > 3: 
+            feature_list.append('SUF_' + data[-3:])
+            
+        feature_list.append('WORD_' + data )
+        
+        return feature_list
+        
+    def tag_sents(self, sents):
+        '''
+        Tag a list of sentences. NB before using this function, user should specify the mode_file either by 
+                       - Train a new model using ``train'' function 
+                       - Use the pre-trained model which is set via ``set_model_file'' function  
+        :params sentences : list of sentences needed to tag. 
+        :type sentences : list(list(str))
+        :return : list of tagged sentences. 
+        :rtype : list (list (tuple(str,str))) 
+        '''
+        tagger = pycrfsuite.Tagger()
+        tagger.open(self._model_file)
+        # We need the list of sentences instead of the list generator for matching the input and output
+        result = []  
+        for sent in sents:
+            features = [self._get_features(data) for data in sent]
+            labels = tagger.tag(features)
+                
+            if len(labels) != len(sent):
+                raise Exception(' Predicted Length Not Matched, Expect Errors !')
+            
+            tagged_sent = [(sent[i],labels[i]) for i in range(len(labels))]
+
+            result.append(tagged_sent)
+            
+        return result 
+    
+    def train(self, train_data, model_file):
+        '''
+        Train the CRF tagger using CRFSuite  
+        :params train_data : is the list of annotated sentences.        
+        :type train_data : list (list(tuple(str,str)))
+        :params model_file : the model will be saved to this file.     
+         
+        '''
+        X_train = []
+        y_train = [] 
+        
+        for sent in train_data:
+            features = [self._get_features(data) for data,label in sent]
+            labels = [label for data,label in sent]
+            
+            X_train.append(features)
+            y_train.append(labels)    
```

Comment:
I think this method can be made smaller if `trainer.append` is called as we go. 'features' is the same as 'xseq', 'labels' is the same as 'yseq', there is no reason to build and unroll X_train and y_train lists.